### ISSUE TITLE
Wrong hinge angles reported for imported UR5 arm

### ISSUE THREAD

**Issue by TylerStephans** [siemens/ros-sharp#330]
* [x] I am at the right place and my issue is directly related to ROS#. General technical questions I would post e.g. at [ROS Answers](https://answers.ros.org/) or [Stack Overflow](https://stackoverflow.com). For library-specific questions I would look for help in the corresponding library forums.
* [x] I have thoroughly read [the Contributing Guideline](Contributing.md) and writing this issue is the right thing to do in my case.

---
## I found a bug! ##
* [x] I am using the latest ROS# version available here on the master branch.
* [x] I am adding all required information, code and data files, screenshots and log files so that you can reproduce the problem.

* My OS is: Windows 10 and Ubuntu 16.04
* My Unity Version is: 2019.4.5f1
* My ROS Distribution is: kinetic
* My Build Target Platform is: Windows and Linux

**Here is my bug description:**
After importing the ur5 arm URDF to Unity, the angles reported by the hinge joints (revolute) are wrong. Specifically, multiple hinges will report changing angles even if only one hinge is actually moving. I've checked that the each child link has the correct parent as a connected body, and they do. I've also tried to recreate the bug by creating my own robot arm by loosely following [this video](https://www.youtube.com/watch?v=07q2mGxD2j8&feature=youtu.be), and angles for its hinge joints are reported correctly.

Perform the following steps to reproduce the bug:

1. Download and build the universal robot package from source by following the instructions on [this page](https://github.com/ros-industrial/universal_robot). Obviously also include the ros-sharp file_server package in your catkin workspace.
2. Launch the rosbridge, and add the robot_description and robot/name to the parameter server.
```
$ roslaunch file_server ros_sharp_communication.launch    
$ roslaunch ur_description ur5_upload.launch transmission_hw_interface:=hardware_interface/EffortJointInterface    
$ rosparam set /robot/name ur5    
```
3. Import the URDF to Unity from ROS using Microsoft serializer and web socket sharp protocol.
4. Enable convex for the ur5 robot, and make the base_link rigidbody kinematic. Also disable use gravity.
5. Check Use Spring in the hinge joints for upper_arm_link down to wrist_3_link, and give a spring constant of 10 and damping of 5 to each.
6. Enable Use Motor for the shoulder_link and set target velocity to -10 and force to 200.

Observed results:
If you check the the Actual Angle reported by the Hinge Joint Limits Manager component of the shoulder_link, you'll see that the angle is changing in magnitude which is expected. However if you look at upper_arm_link or wrist_1_link, these angles are also changing even though they don't appear to be moving. To be clear, I am not referring to small angle changes due to the arm wiggling as it moves. I am observing the angles moving smoothly to -180 deg. and back to 0 again. This leads to incorrect joint states being observed by ROS and the UR5 not being displayed correctly in Rviz.

Expected results:
Because each link has its parent link as the Connected Body in its Hinge Joint, the only hinge that should be observing a meaningful angle change is the shoulder_link.

I'm not 100% certain this issue is because of ros-sharp, but I'm really stumped as to why this is happening or how to fix it? I'd like to avoid manually calculating hinge angles if possible.

**Comment by TylerStephans** [comment]
Interestingly, I've come across [this post](https://forum.unity.com/threads/angle-of-hingejoint-with-connected-body.443101/) on the Unity forums which seems to indicate it is actually a Unity problem. Somehow the angle property of the hinge joint is not necessarily the angle between the two connected bodies...

I think it would be best to leave this issue open for now since it does affect the usability of ros-sharp. Even though I would think it's a Unity issue, the fact that this is still how hinge angles work makes me feel like maybe Unity doesn't consider it a bug? For right now I'll just calculate angles myself, but I'd be interested to hear alternative solutions or what actions if any should be taken to address this :)

**Comment by TylerStephans** [comment]
Thank you @MartinBischoff for your comment. I have gone and posted an example of this issue separate from ROS# [here](https://answers.unity.com/questions/1760939/hingeangle-returns-wrong-angle.html). I've also submitted a bug report, so hopefully this issue will finally be fixed. I have been thinking about adding damping manually as well since the angle issue seems to result in unrealistic torques being applied through the damping included in the hinge joint. I'd like to think that shouldn't cause problems, but there's the other issue of the hinge joint limits being affected which makes my head spin (pun not intended).

Also thank you for your code. I did go and calculate the angle myself, and it seems I did it similarly to yours which is good to know, although mine is less compact.
